modules/server/src/completion.ts

```diff
@@ -0,0 +1,339 @@
+        prefix: context.prefix,
+        linePrefix: context.linePrefix
+      })
+      
+      // Get relevant keywords based on context
+      let relevantKeywords: GosuKeywordInfo[]
+      
+      // First get keywords based on context
+      const contextKeywords = this.getKeywordsForContext(context)
+      this.debug(`Context keywords:`, contextKeywords.map(k => k.keyword))
+      
+      if (context.prefix) {
+        // Filter context keywords by prefix
+        relevantKeywords = contextKeywords.filter(keyword =>
+          keyword.keyword.toLowerCase().startsWith(context.prefix.toLowerCase())
+        )
+        this.debug(`Found ${relevantKeywords.length} keywords matching prefix "${context.prefix}": ${relevantKeywords.map(k => k.keyword)}`)
+      } else {
+        // Use all context keywords
+        relevantKeywords = contextKeywords
+        this.debug(`Found ${relevantKeywords.length} keywords for context`)
+      }
+      
+      // Convert keywords to completion items
+      const keywordCompletions = relevantKeywords.map(keyword =>
+        this.createCompletionItem(keyword)
+      )
+      allCompletions.push(...keywordCompletions)
+      this.debug(`Added ${keywordCompletions.length} keyword completions`)
+      
+      this.debug(`Returning ${allCompletions.length} total completions`)
+      return allCompletions
+      
+    } catch (error) {
+      this.debug('Error getting completions:', error)
+      return []
+    }
+  }
+
+  /**
+   * Analyze the context around the cursor position
+   */
+  private analyzeContext(document: TextDocument, position: Position): CompletionContext {
+    const text = document.getText()
+    const lines = text.split('\n')
+    const currentLine = lines[position.line] || ''
+    const linePrefix = currentLine.substring(0, position.character)
+    
+    // Extract the current word/prefix being typed
+    const wordMatch = linePrefix.match(/(\w+)$/)
+    const prefix = wordMatch ? wordMatch[1] : ''
+    
+    // Analyze document structure
+    const isFileStart = this.isAtFileStart(lines, position.line)
+    const isInClassBody = this.isInClassBody(lines, position.line)
+    const isInFunctionBody = this.isInFunctionBody(lines, position.line)
+    
+    return {
+      isFileStart,
+      isInClassBody,
+      isInFunctionBody,
+      prefix,
+      linePrefix: linePrefix.trim()
+    }
+  }
+
+  /**
+   * Check if position is at the start of the file (only package/imports expected)
+   */
+  private isAtFileStart(lines: string[], currentLine: number): boolean {
+    // Check if we're in the first few lines and no class/interface has been declared yet
+    for (let i = 0; i <= currentLine && i < lines.length; i++) {
+      const line = lines[i].trim()
+      if (line.match(/^(public|private|protected|internal)?\s*(class|interface|enhancement|enum|structure)\s+/)) {
+        return false
+      }
+    }
+    return currentLine < 3 // More restrictive: only first 3 lines are "file start"
```

Comment:
What if we have copyright comments at the beginning of the file? we need a better solution that just a first 3 lines

Reply:
also its a magic number, should be refactored into constant usage